File: runtimes/python/src/twilic/v2.py
"""V2 wire profile encoding and decoding."""

from __future__ import annotations

import struct

from .errors import invalid_data, invalid_tag
from .model import (
    MapEntry,
    Value,
    ValueKind,
    entry,
    new_array,
    new_binary,
    new_bool,
    new_f64,
    new_i64,
    new_map,
    new_null,
    new_string,
    new_u64,
)
from .session import shape_key
from .wire import (
    Reader,
    append_f64_le,
    append_u64_le,
    bounded_decode,
    encode_varuint,
    new_reader,
    read_f64_le,
    read_u64_le,
)

NULL_TAG = 0xC0
FALSE_TAG = 0xC1
TRUE_TAG = 0xC2
F64_TAG = 0xC3
U8_TAG = 0xC4
U16_TAG = 0xC5
U32_TAG = 0xC6
U64_TAG = 0xC7
I8_TAG = 0xC8
I16_TAG = 0xC9
I32_TAG = 0xCA
I64_TAG = 0xCB
BIN8_TAG = 0xCC
BIN16_TAG = 0xCD
BIN32_TAG = 0xCE
STR8_TAG = 0xCF
STR16_TAG = 0xD0
STR32_TAG = 0xD1
ARRAY16_TAG = 0xD2
ARRAY32_TAG = 0xD3
MAP16_TAG = 0xD4
MAP32_TAG = 0xD5
SHAPE_DEF_TAG = 0xD6
KEY_REF_TAG = 0xD8
STR_REF_TAG = 0xD9


class _V2EncodeState:
    __slots__ = ("key_ids", "str_ids", "shape_ids", "next_key_id", "next_str_id", "next_shape_id")

    def __init__(self) -> None:
        self.key_ids: dict[str, int] = {}
        self.str_ids: dict[str, int] = {}
        self.shape_ids: dict[str, int] = {}
        self.next_key_id = 0
        self.next_str_id = 0
        self.next_shape_id = 0


class _V2DecodeState:
    __slots__ = ("keys", "strings", "shapes")

    def __init__(self) -> None:
        self.keys: list[str] = []
        self.strings: list[str] = []
        self.shapes: list[list[str] | None] = []


def encode_v2(value: Value) -> bytes:
    out = bytearray()
    state = _V2EncodeState()
    encode_v2_value(value, out, state)
    return bytes(out)


def decode_v2(data: bytes) -> Value:
    reader = new_reader(data)
    state = _V2DecodeState()
    value = decode_v2_value(reader, state)
    if not reader.is_eof():
        raise invalid_data("trailing bytes in v2 decode")
    return value


def encode_v2_value(value: Value, out: bytearray, state: _V2EncodeState) -> None:
    match value.kind:
        case ValueKind.NULL:
            out.append(NULL_TAG)
        case ValueKind.BOOL:
            out.append(TRUE_TAG if value.bool else FALSE_TAG)
        case ValueKind.I64:
            encode_v2_i64(value.i64, out)
        case ValueKind.U64:
            encode_v2_u64(value.u64, out)
        case ValueKind.F64:
            out.append(F64_TAG)
            append_f64_le(out, value.f64)
        case ValueKind.STRING:
            ref_id = state.str_ids.get(value.str)
            if ref_id is not None:
                out.append(STR_REF_TAG)
                encode_varuint(ref_id, out)
            else:
                encode_v2_string_literal(value.str, out)
                state.str_ids[value.str] = state.next_str_id
                state.next_str_id += 1
        case ValueKind.BINARY:
            encode_v2_binary(value.bin, out)
        case ValueKind.ARRAY:
            encode_v2_array(value.arr, out, state)
        case ValueKind.MAP:
            encode_v2_map(value.map, out, state)
        case _:
            raise invalid_data("unsupported value kind")


def encode_v2_array(values: list[Value], out: bytearray, state: _V2EncodeState) -> None:
    shape_keys = detect_shape_keys(values)
    if shape_keys is not None:
        sk = shape_key(shape_keys)
        shape_id = state.shape_ids.get(sk)
        if shape_id is None:
            shape_id = state.next_shape_id
            state.next_shape_id += 1
            state.shape_ids[sk] = shape_id
        write_v2_array_header(len(values), out)
        out.append(SHAPE_DEF_TAG)
        encode_varuint(shape_id, out)
        encode_varuint(len(shape_keys), out)
        for key in shape_keys:
            encode_v2_key(key, out, state)
        for value in values:
            if value.kind != ValueKind.MAP:
                raise invalid_data("shape array row must be map")
            for field in value.map:
                encode_v2_value(field.value, out, state)
        return
    write_v2_array_header(len(values), out)
    for value in values:
        encode_v2_value(value, out, state)


def encode_v2_map(entries: list[MapEntry], out: bytearray, state: _V2EncodeState) -> None:
    write_v2_map_header(len(entries), out)
    for entry_ in entries:
        encode_v2_key(entry_.key, out, state)
        encode_v2_value(entry_.value, out, state)


def encode_v2_key(key: str, out: bytearray, state: _V2EncodeState) -> None:
    ref_id = state.key_ids.get(key)
    if ref_id is not None:
        out.append(KEY_REF_TAG)
        encode_varuint(ref_id, out)
        return
    encode_v2_string_literal(key, out)
    state.key_ids[key] = state.next_key_id
    state.next_key_id += 1


def encode_v2_string_literal(value: str, out: bytearray) -> None:
    raw = value.encode("utf-8")
    length = len(raw)
    if length <= 31:
        out.append(0x80 | length)
    elif length <= 0xFF:
        out.extend([STR8_TAG, length])
    elif length <= 0xFFFF:
        out.extend([STR16_TAG, length & 0xFF, (length >> 8) & 0xFF])
    else:
        out.append(STR32_TAG)
        out.extend(
            [
                length & 0xFF,
                (length >> 8) & 0xFF,
                (length >> 16) & 0xFF,
                (length >> 24) & 0xFF,
            ]
        )
    out.extend(raw)


def encode_v2_binary(value: bytes, out: bytearray) -> None:
    length = len(value)
    if length <= 0xFF:
        out.extend([BIN8_TAG, length])
    elif length <= 0xFFFF:
        out.extend([BIN16_TAG, length & 0xFF, (length >> 8) & 0xFF])
    else:
        out.append(BIN32_TAG)
        out.extend(
            [
                length & 0xFF,
                (length >> 8) & 0xFF,
                (length >> 16) & 0xFF,
                (length >> 24) & 0xFF,
            ]
        )
    out.extend(value)


def encode_v2_u64(value: int, out: bytearray) -> None:
    if value <= 127:
        out.append(value)
    elif value <= 0xFF:
        out.extend([U8_TAG, value])
    elif value <= 0xFFFF:
        out.extend([U16_TAG, value & 0xFF, (value >> 8) & 0xFF])
    elif value <= 0xFFFFFFFF:
        out.extend(
            [
                U32_TAG,
                value & 0xFF,
                (value >> 8) & 0xFF,
                (value >> 16) & 0xFF,
                (value >> 24) & 0xFF,
            ]
        )
    else:
        out.append(U64_TAG)
        append_u64_le(out, value)


def encode_v2_i64(value: int, out: bytearray) -> None:
    if -32 <= value <= -1:
        out.append(value & 0xFF)
    elif 0 <= value <= 127:
        out.append(value)
    elif -128 <= value <= 127:
        out.extend([I8_TAG, value & 0xFF])
    elif -32768 <= value <= 32767:
        out.extend([I16_TAG, value & 0xFF, (value >> 8) & 0xFF])
    elif -2147483648 <= value <= 2147483647:
        out.extend(
            [
                I32_TAG,
                value & 0xFF,
                (value >> 8) & 0xFF,
                (value >> 16) & 0xFF,
                (value >> 24) & 0xFF,
            ]
        )
    else:
        out.append(I64_TAG)
        append_u64_le(out, value & 0xFFFFFFFFFFFFFFFF)


def write_v2_array_header(length: int, out: bytearray) -> None:
    if length <= 15:
        out.append(0xA0 | length)
    elif length <= 0xFFFF:
        out.extend([ARRAY16_TAG, length & 0xFF, (length >> 8) & 0xFF])
    else:
        out.append(ARRAY32_TAG)
        out.extend(
            [
                length & 0xFF,
                (length >> 8) & 0xFF,
                (length >> 16) & 0xFF,
                (length >> 24) & 0xFF,
            ]
        )


def write_v2_map_header(length: int, out: bytearray) -> None:
    if length <= 15:
        out.append(0xB0 | length)
    elif length <= 0xFFFF:
        out.extend([MAP16_TAG, length & 0xFF, (length >> 8) & 0xFF])
    else:
        out.append(MAP32_TAG)
        out.extend(
            [
                length & 0xFF,
                (length >> 8) & 0xFF,
                (length >> 16) & 0xFF,
                (length >> 24) & 0xFF,
            ]
        )


def detect_shape_keys(values: list[Value]) -> list[str] | None:
    if len(values) < 2:
        return None
    if values[0].kind != ValueKind.MAP or not values[0].map:
        return None
    keys = [e.key for e in values[0].map]
    for value in values[1:]:
        if value.kind != ValueKind.MAP or len(value.map) != len(keys):
            return None
        for i, e in enumerate(value.map):
            if e.key != keys[i]:
                return None
    return keys


def decode_v2_value(reader: Reader, state: _V2DecodeState) -> Value:
    tag = reader.read_u8()
    return decode_v2_value_from_tag(reader, state, tag)


def decode_v2_value_from_tag(reader: Reader, state: _V2DecodeState, tag: int) -> Value:
    if tag <= 0x7F:
        return new_u64(tag)
    if 0x80 <= tag <= 0x9F:
        length = tag & 0x1F
        raw = reader.read_exact(length)
        s = raw.decode("utf-8")
        state.strings.append(s)
        return new_string(s)
    if 0xA0 <= tag <= 0xAF:
        return decode_v2_array_body(reader, state, tag & 0x0F)
    if 0xB0 <= tag <= 0xBF:
        return decode_v2_map_body(reader, state, tag & 0x0F)
    if tag >= 0xE0:
        return new_i64(tag if tag < 128 else tag - 256)
    if tag == NULL_TAG:
        return new_null()
    if tag == FALSE_TAG:
        return new_bool(False)
    if tag == TRUE_TAG:
        return new_bool(True)
    if tag == F64_TAG:
        return new_f64(read_f64_le(reader))
    if tag == U8_TAG:
        return new_u64(reader.read_u8())
    if tag == U16_TAG:
        b = reader.read_exact(2)
        return new_u64(b[0] | (b[1] << 8))
    if tag == U32_TAG:
        b = reader.read_exact(4)
        return new_u64(b[0] | (b[1] << 8) | (b[2] << 16) | (b[3] << 24))
    if tag == U64_TAG:
        return new_u64(read_u64_le(reader))
    if tag == I8_TAG:
        b = reader.read_u8()
        return new_i64(b if b < 128 else b - 256)
    if tag == I16_TAG:
        b = reader.read_exact(2)
        return new_i64(struct.unpack("<h", b)[0])
    if tag == I32_TAG:
        b = reader.read_exact(4)
        return new_i64(struct.unpack("<i", b)[0])
    if tag == I64_TAG:
        b = reader.read_exact(8)
        return new_i64(struct.unpack("<q", b)[0])
    if tag == BIN8_TAG:
        n = reader.read_u8()
        return new_binary(reader.read_exact(n))
    if tag == BIN16_TAG:
        b = reader.read_exact(2)
        n = b[0] | (b[1] << 8)
        return new_binary(reader.read_exact(n))
    if tag == BIN32_TAG:
        b = reader.read_exact(4)
        n = b[0] | (b[1] << 8) | (b[2] << 16) | (b[3] << 24)
        return new_binary(reader.read_exact(n))
    if tag in (STR8_TAG, STR16_TAG, STR32_TAG):
        return decode_v2_string_tag(reader, state, tag)
    if tag == ARRAY16_TAG:
        b = reader.read_exact(2)
        n = b[0] | (b[1] << 8)
        return decode_v2_array_body(reader, state, n)
    if tag == ARRAY32_TAG:
        b = reader.read_exact(4)
        n = b[0] | (b[1] << 8) | (b[2] << 16) | (b[3] << 24)
        return decode_v2_array_body(reader, state, n)
    if tag == MAP16_TAG:
        b = reader.read_exact(2)
        n = b[0] | (b[1] << 8)
        return decode_v2_map_body(reader, state, n)
    if tag == MAP32_TAG:
        b = reader.read_exact(4)
        n = b[0] | (b[1] << 8) | (b[2] << 16) | (b[3] << 24)
        return decode_v2_map_body(reader, state, n)
    if tag == STR_REF_TAG:
        ref_id = reader.read_varuint()
        if ref_id >= len(state.strings):
            raise invalid_data("unknown str_ref id")
        return new_string(state.strings[ref_id])
    raise invalid_tag(tag)


def decode_v2_string_tag(reader: Reader, state: _V2DecodeState, tag: int) -> Value:
    if tag == STR8_TAG:
        length = reader.read_u8()
    elif tag == STR16_TAG:
        b = reader.read_exact(2)
        length = b[0] | (b[1] << 8)
    elif tag == STR32_TAG:
        b = reader.read_exact(4)
        length = b[0] | (b[1] << 8) | (b[2] << 16) | (b[3] << 24)
    else:
        raise invalid_data("invalid string tag")
    raw = reader.read_exact(length)
    s = raw.decode("utf-8")
    state.strings.append(s)
    return new_string(s)


@bounded_decode
def decode_v2_array_body(reader: Reader, state: _V2DecodeState, length: int) -> Value:
    reader.claim_output(length)
    if length == 0:
        return new_array([])
    first_tag = reader.read_u8()
    if first_tag == SHAPE_DEF_TAG:
        shape_id = reader.read_count(65_535)
        key_count = reader.read_count(256)
        keys: list[str] = []
        for _ in range(key_count):
            keys.append(decode_v2_key(reader, state))
        while shape_id >= len(state.shapes):
            state.shapes.append(None)
        state.shapes[shape_id] = keys
        values: list[Value] = []
        for _ in range(length):
            reader.claim_output(key_count)
            row = [entry(key, decode_v2_value(reader, state)) for key in keys]
            values.append(new_map(*row))
        return new_array(values)
    values = [None] * length  # type: ignore[list-item]
    values[0] = decode_v2_value_from_tag(reader, state, first_tag)
    for i in range(1, length):
        values[i] = decode_v2_value(reader, state)
    return new_array(values)  # type: ignore[arg-type]


@bounded_decode
def decode_v2_map_body(reader: Reader, state: _V2DecodeState, length: int) -> Value:
    reader.claim_output(length)
    entries: list[MapEntry] = []
    for _ in range(length):
        key = decode_v2_key(reader, state)
        value = decode_v2_value(reader, state)
        entries.append(entry(key, value))
    return new_map(*entries)


def decode_v2_key(reader: Reader, state: _V2DecodeState) -> str:
    tag = reader.read_u8()
    if tag == KEY_REF_TAG:
        ref_id = reader.read_varuint()
        if ref_id >= len(state.keys):
            raise invalid_data("unknown key_ref id")
        return state.keys[ref_id]
    if 0x80 <= tag <= 0x9F:
        length = tag & 0x1F
        key = reader.read_exact(length).decode("utf-8")
        state.keys.append(key)
        return key
    if tag in (STR8_TAG, STR16_TAG, STR32_TAG):
        v = decode_v2_value_from_tag(reader, state, tag)
        if v.kind != ValueKind.STRING:
            raise invalid_data("expected string key")
        state.keys.append(v.str)
        return v.str
    raise invalid_data("map key must be key_ref or string")

File: runtimes/dart/lib/src/v2.dart
import 'dart:convert';
import 'dart:typed_data';

import 'errors.dart';
import 'model.dart';
import 'session.dart' show shapeKey;
import 'wire.dart';

const _nullTag = 0xC0;
const _falseTag = 0xC1;
const _trueTag = 0xC2;
const _f64Tag = 0xC3;
const _u8Tag = 0xC4;
const _u16Tag = 0xC5;
const _u32Tag = 0xC6;
const _u64Tag = 0xC7;
const _i8Tag = 0xC8;
const _i16Tag = 0xC9;
const _i32Tag = 0xCA;
const _i64Tag = 0xCB;
const _bin8Tag = 0xCC;
const _bin16Tag = 0xCD;
const _bin32Tag = 0xCE;
const _str8Tag = 0xCF;
const _str16Tag = 0xD0;
const _str32Tag = 0xD1;
const _array16Tag = 0xD2;
const _array32Tag = 0xD3;
const _map16Tag = 0xD4;
const _map32Tag = 0xD5;
const _shapeDefTag = 0xD6;
const _keyRefTag = 0xD8;
const _strRefTag = 0xD9;

class _V2EncodeState {
  final keyIds = <String, int>{};
  final strIds = <String, int>{};
  final shapeIds = <String, int>{};
  int nextKeyId = 0;
  int nextStrId = 0;
  int nextShapeId = 0;
}

class _V2DecodeState {
  final keys = <String>[];
  final strings = <String>[];
  final shapes = <List<String>?>[];
}

Uint8List encodeV2(Value value) {
  final out = BytesBuilder();
  final state = _V2EncodeState();
  _encodeV2Value(value, out, state);
  return out.toBytes();
}

Value decodeV2(Uint8List data) {
  final reader = newReader(data);
  final state = _V2DecodeState();
  final value = _decodeV2Value(reader, state);
  if (!reader.isEof) throw invalidData('trailing bytes in v2 decode');
  return value;
}

void _encodeV2Value(Value value, BytesBuilder out, _V2EncodeState state) {
  switch (value.kind) {
    case ValueKind.nullKind:
      out.addByte(_nullTag);
    case ValueKind.boolKind:
      out.addByte(value.boolValue ? _trueTag : _falseTag);
    case ValueKind.i64:
      _encodeV2I64(value.i64, out);
    case ValueKind.u64:
      _encodeV2U64(value.u64, out);
    case ValueKind.f64:
      out.addByte(_f64Tag);
      appendF64Le(out, value.f64);
    case ValueKind.string:
      final refId = state.strIds[value.str];
      if (refId != null) {
        out.addByte(_strRefTag);
        encodeVaruint(refId, out);
      } else {
        _encodeV2StringLiteral(value.str, out);
        state.strIds[value.str] = state.nextStrId++;
      }
    case ValueKind.binary:
      _encodeV2Binary(value.bin, out);
    case ValueKind.array:
      _encodeV2Array(value.arr, out, state);
    case ValueKind.map:
      _encodeV2Map(value.map, out, state);
  }
}

void _encodeV2Array(
    List<Value> values, BytesBuilder out, _V2EncodeState state) {
  final shapeKeys = _detectShapeKeys(values);
  if (shapeKeys != null) {
    final sk = shapeKey(shapeKeys);
    var shapeId = state.shapeIds[sk];
    if (shapeId == null) {
      shapeId = state.nextShapeId++;
      state.shapeIds[sk] = shapeId;
    }
    _writeV2ArrayHeader(values.length, out);
    out.addByte(_shapeDefTag);
    encodeVaruint(shapeId, out);
    encodeVaruint(shapeKeys.length, out);
    for (final key in shapeKeys) {
      _encodeV2Key(key, out, state);
    }
    for (final value in values) {
      if (value.kind != ValueKind.map) {
        throw invalidData('shape array row must be map');
      }
      for (final field in value.map) {
        _encodeV2Value(field.value, out, state);
      }
    }
    return;
  }
  _writeV2ArrayHeader(values.length, out);
  for (final value in values) {
    _encodeV2Value(value, out, state);
  }
}

void _encodeV2Map(
    List<MapEntry> entries, BytesBuilder out, _V2EncodeState state) {
  _writeV2MapHeader(entries.length, out);
  for (final e in entries) {
    _encodeV2Key(e.key, out, state);
    _encodeV2Value(e.value, out, state);
  }
}

void _encodeV2Key(String key, BytesBuilder out, _V2EncodeState state) {
  final refId = state.keyIds[key];
  if (refId != null) {
    out.addByte(_keyRefTag);
    encodeVaruint(refId, out);
    return;
  }
  _encodeV2StringLiteral(key, out);
  state.keyIds[key] = state.nextKeyId++;
}

void _encodeV2StringLiteral(String value, BytesBuilder out) {
  final raw = utf8.encode(value);
  final length = raw.length;
  if (length <= 31) {
    out.addByte(0x80 | length);
  } else if (length <= 0xFF) {
    out.addByte(_str8Tag);
    out.addByte(length);
  } else if (length <= 0xFFFF) {
    out.addByte(_str16Tag);
    out.addByte(length & 0xFF);
    out.addByte((length >> 8) & 0xFF);
  } else {
    out.addByte(_str32Tag);
    out.addByte(length & 0xFF);
    out.addByte((length >> 8) & 0xFF);
    out.addByte((length >> 16) & 0xFF);
    out.addByte((length >> 24) & 0xFF);
  }
  out.add(raw);
}

void _encodeV2Binary(Uint8List value, BytesBuilder out) {
  final length = value.length;
  if (length <= 0xFF) {
    out.addByte(_bin8Tag);
    out.addByte(length);
  } else if (length <= 0xFFFF) {
    out.addByte(_bin16Tag);
    out.addByte(length & 0xFF);
    out.addByte((length >> 8) & 0xFF);
  } else {
    out.addByte(_bin32Tag);
    out.addByte(length & 0xFF);
    out.addByte((length >> 8) & 0xFF);
    out.addByte((length >> 16) & 0xFF);
    out.addByte((length >> 24) & 0xFF);
  }
  out.add(value);
}

void _encodeV2U64(int value, BytesBuilder out) {
  if (value <= 127) {
    out.addByte(value);
  } else if (value <= 0xFF) {
    out.addByte(_u8Tag);
    out.addByte(value);
  } else if (value <= 0xFFFF) {
    out.addByte(_u16Tag);
    out.addByte(value & 0xFF);
    out.addByte((value >> 8) & 0xFF);
  } else if (value <= 0xFFFFFFFF) {
    out.addByte(_u32Tag);
    out.addByte(value & 0xFF);
    out.addByte((value >> 8) & 0xFF);
    out.addByte((value >> 16) & 0xFF);
    out.addByte((value >> 24) & 0xFF);
  } else {
    out.addByte(_u64Tag);
    appendU64Le(out, value);
  }
}

void _encodeV2I64(int value, BytesBuilder out) {
  if (value >= -32 && value <= -1) {
    out.addByte(value & 0xFF);
  } else if (value >= 0 && value <= 127) {
    out.addByte(value);
  } else if (value >= -128 && value <= 127) {
    out.addByte(_i8Tag);
    out.addByte(value & 0xFF);
  } else if (value >= -32768 && value <= 32767) {
    out.addByte(_i16Tag);
    out.addByte(value & 0xFF);
    out.addByte((value >> 8) & 0xFF);
  } else if (value >= -2147483648 && value <= 2147483647) {
    out.addByte(_i32Tag);
    out.addByte(value & 0xFF);
    out.addByte((value >> 8) & 0xFF);
    out.addByte((value >> 16) & 0xFF);
    out.addByte((value >> 24) & 0xFF);
  } else {
    out.addByte(_i64Tag);
    appendU64Le(out, value);
  }
}

void _writeV2ArrayHeader(int length, BytesBuilder out) {
  if (length <= 15) {
    out.addByte(0xA0 | length);
  } else if (length <= 0xFFFF) {
    out.addByte(_array16Tag);
    out.addByte(length & 0xFF);
    out.addByte((length >> 8) & 0xFF);
  } else {
    out.addByte(_array32Tag);
    out.addByte(length & 0xFF);
    out.addByte((length >> 8) & 0xFF);
    out.addByte((length >> 16) & 0xFF);
    out.addByte((length >> 24) & 0xFF);
  }
}

void _writeV2MapHeader(int length, BytesBuilder out) {
  if (length <= 15) {
    out.addByte(0xB0 | length);
  } else if (length <= 0xFFFF) {
    out.addByte(_map16Tag);
    out.addByte(length & 0xFF);
    out.addByte((length >> 8) & 0xFF);
  } else {
    out.addByte(_map32Tag);
    out.addByte(length & 0xFF);
    out.addByte((length >> 8) & 0xFF);
    out.addByte((length >> 16) & 0xFF);
    out.addByte((length >> 24) & 0xFF);
  }
}

List<String>? _detectShapeKeys(List<Value> values) {
  if (values.length < 2) return null;
  if (values[0].kind != ValueKind.map || values[0].map.isEmpty) return null;
  final keys = values[0].map.map((e) => e.key).toList();
  for (var i = 1; i < values.length; i++) {
    final value = values[i];
    if (value.kind != ValueKind.map || value.map.length != keys.length)
      return null;
    for (var j = 0; j < keys.length; j++) {
      if (value.map[j].key != keys[j]) return null;
    }
  }
  return keys;
}

Value _decodeV2Value(Reader reader, _V2DecodeState state) {
  final tag = reader.readU8();
  return _decodeV2ValueFromTag(reader, state, tag);
}

Value _decodeV2ValueFromTag(Reader reader, _V2DecodeState state, int tag) {
  if (tag <= 0x7F) return newU64(tag);
  if (tag >= 0x80 && tag <= 0x9F) {
    final length = tag & 0x1F;
    final s = utf8.decode(reader.readExact(length));
    state.strings.add(s);
    return newString(s);
  }
  if (tag >= 0xA0 && tag <= 0xAF) {
    return _decodeV2ArrayBody(reader, state, tag & 0x0F);
  }
  if (tag >= 0xB0 && tag <= 0xBF) {
    return _decodeV2MapBody(reader, state, tag & 0x0F);
  }
  if (tag >= 0xE0) return newI64(tag < 128 ? tag : tag - 256);
  switch (tag) {
    case _nullTag:
      return newNull();
    case _falseTag:
      return newBool(false);
    case _trueTag:
      return newBool(true);
    case _f64Tag:
      return newF64(readF64Le(reader));
    case _u8Tag:
      return newU64(reader.readU8());
    case _u16Tag:
      final b = reader.readExact(2);
      return newU64(b[0] | (b[1] << 8));
    case _u32Tag:
      final b = reader.readExact(4);
      return newU64(b[0] | (b[1] << 8) | (b[2] << 16) | (b[3] << 24));
    case _u64Tag:
      return newU64(readU64Le(reader));
    case _i8Tag:
      final b = reader.readU8();
      return newI64(b < 128 ? b : b - 256);
    case _i16Tag:
      final b = reader.readExact(2);
      final bd = ByteData.sublistView(b);
      return newI64(bd.getInt16(0, Endian.little));
    case _i32Tag:
      final b = reader.readExact(4);
      final bd = ByteData.sublistView(b);
      return newI64(bd.getInt32(0, Endian.little));
    case _i64Tag:
      final b = reader.readExact(8);
      final bd = ByteData.sublistView(b);
      return newI64(bd.getInt64(0, Endian.little));
    case _bin8Tag:
      return newBinary(reader.readExact(reader.readU8()));
    case _bin16Tag:
      final b = reader.readExact(2);
      return newBinary(reader.readExact(b[0] | (b[1] << 8)));
    case _bin32Tag:
      final b = reader.readExact(4);
      final n = b[0] | (b[1] << 8) | (b[2] << 16) | (b[3] << 24);
      return newBinary(reader.readExact(n));
    case _str8Tag:
    case _str16Tag:
    case _str32Tag:
      return _decodeV2StringTag(reader, state, tag);
    case _array16Tag:
      final b = reader.readExact(2);
      return _decodeV2ArrayBody(reader, state, b[0] | (b[1] << 8));
    case _array32Tag:
      final b = reader.readExact(4);
      final n = b[0] | (b[1] << 8) | (b[2] << 16) | (b[3] << 24);
      return _decodeV2ArrayBody(reader, state, n);
    case _map16Tag:
      final b = reader.readExact(2);
      return _decodeV2MapBody(reader, state, b[0] | (b[1] << 8));
    case _map32Tag:
      final b = reader.readExact(4);
      final n = b[0] | (b[1] << 8) | (b[2] << 16) | (b[3] << 24);
      return _decodeV2MapBody(reader, state, n);
    case _strRefTag:
      final refId = reader.readVaruint();
      if (refId >= state.strings.length)
        throw invalidData('unknown str_ref id');
      return newString(state.strings[refId]);
    default:
      throw invalidTag(tag);
  }
}

Value _decodeV2StringTag(Reader reader, _V2DecodeState state, int tag) {
  late final int length;
  if (tag == _str8Tag) {
    length = reader.readU8();
  } else if (tag == _str16Tag) {
    final b = reader.readExact(2);
    length = b[0] | (b[1] << 8);
  } else {
    final b = reader.readExact(4);
    length = b[0] | (b[1] << 8) | (b[2] << 16) | (b[3] << 24);
  }
  final s = utf8.decode(reader.readExact(length));
  state.strings.add(s);
  return newString(s);
}

Value _decodeV2ArrayBody(Reader reader, _V2DecodeState state, int length) {
  reader.claimOutput(length);
  return reader.withDepth(() => _decodeV2ArrayBodyInner(reader, state, length));
}

Value _decodeV2ArrayBodyInner(Reader reader, _V2DecodeState state, int length) {
  if (length == 0) return newArray([]);
  final firstTag = reader.readU8();
  if (firstTag == _shapeDefTag) {
    final shapeId = reader.readCount(65535);
    final keyCount = reader.readCount(256);
    final keys = <String>[];
    for (var i = 0; i < keyCount; i++) {
      keys.add(_decodeV2Key(reader, state));
    }
    while (state.shapes.length <= shapeId) {
      state.shapes.add(null);
    }
    state.shapes[shapeId] = keys;
    final values = <Value>[];
    for (var i = 0; i < length; i++) {
      reader.claimOutput(keys.length);
      final row = <MapEntry>[];
      for (final key in keys) {
        row.add(MapEntry(key, _decodeV2Value(reader, state)));
      }
      values.add(newMap(row));
    }
    return newArray(values);
  }
  final values = <Value>[_decodeV2ValueFromTag(reader, state, firstTag)];
  for (var i = 1; i < length; i++) {
    values.add(_decodeV2Value(reader, state));
  }
  return newArray(values);
}

Value _decodeV2MapBody(Reader reader, _V2DecodeState state, int length) {
  reader.claimOutput(length);
  return reader.withDepth(() => _decodeV2MapBodyInner(reader, state, length));
}

Value _decodeV2MapBodyInner(Reader reader, _V2DecodeState state, int length) {
  final entries = <MapEntry>[];
  for (var i = 0; i < length; i++) {
    entries.add(
        MapEntry(_decodeV2Key(reader, state), _decodeV2Value(reader, state)));
  }
  return newMap(entries);
}

String _decodeV2Key(Reader reader, _V2DecodeState state) {
  final tag = reader.readU8();
  if (tag == _keyRefTag) {
    final refId = reader.readVaruint();
    if (refId >= state.keys.length) throw invalidData('unknown key_ref id');
    return state.keys[refId];
  }
  if (tag >= 0x80 && tag <= 0x9F) {
    final key = utf8.decode(reader.readExact(tag & 0x1F));
    state.keys.add(key);
    return key;
  }
  if (tag == _str8Tag || tag == _str16Tag || tag == _str32Tag) {
    final v = _decodeV2ValueFromTag(reader, state, tag);
    if (v.kind != ValueKind.string) throw invalidData('expected string key');
    state.keys.add(v.str);
    return v.str;
  }
  throw invalidData('map key must be key_ref or string');
}
